compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -158,6 +158,8 @@ pub struct ResolverOutputs {
 #[derive(Debug)]
 pub struct ResolverGlobalCtxt {
     pub visibilities: FxHashMap<LocalDefId, Visibility>,
+    pub impl_restrictions: FxHashMap<DefId, Restriction>,
+    pub mut_restrictions: FxHashMap<DefId, Restriction>,
```

Comment:
Could those have different types to make sure we don't use one instead of another?

Reply:
I'm looking into this, but it's quite a bit more involved that it initially seems.

Reply:
Managed to work around it. It's a bit hacky, but I'm comfortable with it.